physics/GFS_MP_generic.F90

```diff
@@ -42,12 +42,24 @@ subroutine GFS_MP_generic_pre_run(im, levs, ldiag3d, do_aw, ntcw, nncl, ntrac, g
         do k=1,levs
           do i=1,im
             save_t(i,k) = gt0(i,k)
-            save_q(1:im,:,1) = gq0(1:im,:,1)
           enddo
         enddo
-        do n=ntcw,ntcw+nncl-1
-          save_q(1:im,:,n) = gq0(1:im,:,n)
-        enddo
+        if(qdiag3d) then
+           do k=1,levs
+              do i=1,im
+                 ! Here, gq0(...,1) is used instead of gq0_water_vapor
+                 ! to be consistent with the GFS_MP_generic_post_run
+                 ! code.
+                 save_qv(i,k) = gq0(i,k,1)
```

Comment:
All things equal, I'd prefer to rely on passing in gq0_water_vapor instead of the whole array and hard-coding the water vapor index. I see that the whole array is used below for do_aw, in which case, OK, no need to pass in the data twice.

Reply:
Instead of `gq0(i,k,1)` we should be using `gq0(i,k,ntqv)` in this case.

Reply:
Thanks @climbfuji. I couldn't remember if there was a named index for water vapor or not. Since there is, as long as the whole tracer array is needed for do_aw, using the named index would definitely be the preferred solution.